driver/maindriver.ml

```diff
@@ -19,96 +19,107 @@ let usage = "Usage: ocamlc <options> <files>\nOptions are:"
-      let revd = Compenv.get_objfiles ~with_ocamlparam:false in
-      Compmisc.with_ppf_dump ~file_prefix:extracted_output (fun ppf_dump ->
+    Bytelibrarian.create_archive
+      (Compenv.get_objfiles ~with_ocamlparam:false)
+      (Compenv.extract_output !output_name);
+    Warnings.check_fatal ();
+  end
+  else if !make_package then begin
+    Compmisc.init_path ();
+    let extracted_output = Compenv.extract_output !output_name in
+    let revd = Compenv.get_objfiles ~with_ocamlparam:false in
+    Compmisc.with_ppf_dump ~file_prefix:extracted_output (fun ppf_dump ->
         Bytepackager.package_files ~ppf_dump (Compmisc.initial_env ())
           revd (extracted_output));
-      Warnings.check_fatal ();
-    end
-    else if not !Compenv.stop_early && !objfiles <> [] then begin
-      let target =
-        if !output_c_object && not !output_complete_executable then
-          let s = Compenv.extract_output !output_name in
-          if (Filename.check_suffix s Config.ext_obj
+    Warnings.check_fatal ();
+  end
+  else if not !Compenv.stop_early && !objfiles <> [] then begin
+    let target =
+      if !output_c_object && not !output_complete_executable then
+        let s = Compenv.extract_output !output_name in
+        if (Filename.check_suffix s Config.ext_obj
             || Filename.check_suffix s Config.ext_dll
             || Filename.check_suffix s ".c")
-          then s
-          else
-            Compenv.fatal
-              (Printf.sprintf
-                 "The extension of the output file must be .c, %s or %s"
-                 Config.ext_obj Config.ext_dll
-              )
+        then s
         else
-          Compenv.default_output !output_name
-      in
-      Compmisc.init_path ();
-      Bytelink.link (Compenv.get_objfiles ~with_ocamlparam:true) target;
-      Warnings.check_fatal ();
-    end;
-  with
-  | exception (Compenv.Exit_with_status n) ->
-    n
-  | exception x ->
-    Location.report_exception ppf x;
-    2
+          Compenv.fatal
+            (Printf.sprintf
+               "The extension of the output file must be .c, %s or %s"
+               Config.ext_obj Config.ext_dll
+            )
+      else
+        Compenv.default_output !output_name
+    in
+    Compmisc.init_path ();
+    Bytelink.link (Compenv.get_objfiles ~with_ocamlparam:true) target;
+    Warnings.check_fatal ();
+  end
+
+let main argv ppf =
+  match process_arguments argv ppf with
+  | exception (Compenv.Exit_with_status n) -> n
   | () ->
-    Profile.print Format.std_formatter !Clflags.profile_columns;
-    0
+      let log = Location.init_log ppf in
+      match process_main log with
+      | exception (Compenv.Exit_with_status n) ->
+          Misc.Log.flush log;
+          n
+      | exception x ->
+          Location.report_exception log x;
+          Misc.Log.flush log;
+          2
+      | () ->
+          Profile.print Format.std_formatter !Clflags.profile_columns;
+          Misc.Log.flush log;
```

Comment:
maybe the three calls to `flush` could be factored here (after the exit code is computed, or as a `try_finally` around the `let log`.)